app/services/fact_cleaner.rb

```diff
@@ -0,0 +1,44 @@
+class FactCleaner
+  delegate :logger, :to => :Rails
+
+  def clean!
+    @to_delete = []
+    logger.info "Searching for orphaned facts..."
+    find_leaf_orphaned_facts
+
+    # we also delete all composes that either don't have any descendant or all of them are to be deleted
+    logger.info "Searching for compose orphaned facts..."
+    find_compose_orphaned_facts
+
+    logger.info "#{@to_delete.size} facts will be removed"
+
+    delete!
+    logger.info "Cleanup of facts finished!"
+    @to_delete.count
+  end
+
+  private
+
+  def find_leaf_orphaned_facts
+    FactName.where(:compose => false).includes(:fact_values).find_in_batches do |batch|
+      @to_delete += batch.select do |fact|
+        fact.fact_values.empty?
+      end
+    end
+  end
+
+  def find_compose_orphaned_facts
```

Comment:
the fact we search for it in batches is specific to this task that is expect to deal with big number of records, if we only extracted the condition in form of scope, then argument from above applies (not worth at least until left outer joins)